spec/compiler/semantic/if_spec.cr

```diff
@@ -337,4 +337,18 @@ describe "Semantic: if" do
       foo
       )) { nilable int32 }
   end
+
+  it "doesn't consider nil type in else branch with if with && (#7434)" do
+    assert_type(%(
+      def foo
+        if true && (y = 1 || nil)
+        else
+          return 1
+        end
+        y
+      end
+
+      foo
+      ), inject_primitives: false) { int32 }
```

Comment:
I don't understand your test, in all cases the return type can only be int32, I think you want to `return false` or some other type?

Reply:
That's the point: This test ensures that the compiler recognizes that the return type is always `Int32`, because the `if` branch can never be reached.

Reply:
oooh ok I think I get it, it'd be far easier to understand if we could do:
```cr
if true && (y = 1 || nil)
end
assert_type(y, Int32)
```
Or maybe just do:
```cr
assert_type(%(
  if true && (y = 1 || nil)
  end
  y
  ), inject_primitives: false) { int32 }
```
?

Reply:
But in your examples y ends up nilable. It's the return in the else branch that makes it non nilable. But that what was not working well (when in a right side of &&).